### ISSUE TITLE
merge.concatenate layer not connected to input layer. Summary not printing correct total number of weights.

### ISSUE THREAD

**Issue by johnmcgowan1** [tensorflow/tensorflow#38817]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): 
- TensorFlow version (use command below): 2.1.0
- Python version: 3.6.8
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

I am working on training a classifier to classify two distributions with an adversary that learns the noise from the distributions. Model.summary() indicates that my adversary model is not connected to input. merge.concatenate layer seems to break the models continuity but I do not understand why.  Loss is not decreasing when I train this model. The combined model consists of 2 models (classifier and adversary) with 501 and 1276 weights respectively. Total number of trainable weights listed by model.summary is 1276 instead of 1777. 

**Describe the expected behavior**

Loss should decrease during training, combined model should have 1777 weights and adversary model should have an entry in summary.connected to.

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

N = 62500

# Make data - 2 class that are 2D gaussian distributed. Gaussian noise affecting y coord of Y=1 class
cov0 = np.array([[1,-0.5],[-0.5,1]])
cov1 = np.array([[1,0],[0,1]])

train_data0 = np.random.multivariate_normal([0,0],cov0,N)
train_labels0 = np.zeros((N,1))
train_data1 = np.random.multivariate_normal([1,1],cov1,N)
train_labels1 = np.ones((N,1))
train_z = np.random.normal(0,1,N*2)
train_data1[:,1] = train_data1[:,1]+train_z[N:]
#train_z[:N] = 0
train_data = np.vstack([train_data0,train_data1])
train_labels = np.append(train_labels0,train_labels1)

#shuffle data
indices = np.random.permutation(len(train_data))
train_data = train_data[indices]
train_labels =train_labels[indices]
train_z = train_z[indices]

#train test split
train_data, valid_data, train_labels, valid_labels, train_z, valid_z = train_test_split(train_data,train_labels,train_z,test_size=50000)

#Models 
inputs = Input(shape=(train_data.shape[1],))

# Classifier model
Dx = Dense(20, activation="tanh")(inputs)
Dx = Dense(20, activation="relu")(Dx)
Dx = Dense(1, activation="sigmoid")(Dx)
f = Model([inputs],[Dx])

#adversary model
from keras.layers.merge import concatenate
out = f(inputs)
out = Dense(20,activation="relu")(out)
out = Dense(20,activation="relu")(out)
mu = Dense(5,activation="linear")(out)
sigma = Dense(5,activation=backend.exp)(out)
pi = Dense(5,activation="softmax")(out)
out = concatenate([mu,sigma,pi])
r = Model([inputs],[out]) 

# Loss for r 
def rloss(lam):
  def loss_r(y_true,y_pred):
    #y_true = y_true.flatten()
    mu = y_pred[:,:5]
    sig = y_pred[:,5:10]
    pi = y_pred[:,10:]
    pdf = 0
    for i in range(5):
      pdf += pi[:,i] *(1.0/(sig[:,i]*np.sqrt(2.0*np.pi)))*backend.exp(-((y_true-mu[:,i])**2)/(2.0*(sig[:,i]**2)))
    return lam * tf.math.reduce_mean(-backend.log(pdf))
  return loss_r

f.compile(loss = backend.binary_crossentropy, optimizer = SGD())

#combined model
adv1 = Model([inputs], [f(inputs), r(inputs)])
adv1.compile(loss=[backend.binary_crossentropy,rloss(-50)],optimizer=SGD(momentum=0.0))

# adversary
adv2 = Model(inputs,[r(inputs)])
adv2.compile(loss=[rloss(1.0)],optimizer=SGD(momentum=0.0))

# adv1 summary total trainable params != sum of params of individual models
adv1.summary()

#adv2 missing "connected to"
adv2.summary()

#Loss of adversary does not decrease as expected
f.trainable = False
r.trainable = True
adv2.fit(train_data,train_z,epochs=20)

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

I am trying to reproduce the model of the paper. Paper's code is on github: https://github.com/glouppe/paper-learning-to-pivot/blob/master/code/Toy.ipynb

**Comment by amahendrakar** [comment]
@johnmcgowan1,
Importing `keras` directly instead of `tensorflow.keras` seems to work. I was able to run the code without any issues, please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/253ed225016ac2afcd15736146df4463/38817.ipynb). 

You can read more about it, from [this StackOverflow](https://stackoverflow.com/a/51831434) comment. Thanks!

**Comment by johnmcgowan1** [comment]
@amahendrakar Thank you for your attention. I have changed to keras but this does not solve the problem Note in rloss that you should be using (-keras.backend.log), in the colab you sent you use keras.backend.log. If you change this to negative you will see the loss still hardly decreases at all. I still expect the number of weights in the first summary to be 1777 and to have a connected comment in summary.

**Comment by johnmcgowan1** [comment]
I was able to solve this by setting KERAS_BACKEND to theano. And compiling the adversary again after setting weights to trainable. Still only 1276 weights are printed, but the loss is decreasing as it should be. Functioning code here:

import numpy as np
import os
import theano
os.environ['KERAS_BACKEND'] ='theano'
import keras
from keras.layers import Input, Dense, Concatenate
from keras import initializers, backend
from keras.models import Model, Sequential
from keras.optimizers import SGD
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split

N = 62500

# Make data
cov0 = np.array([[1,-0.5],[-0.5,1]])
cov1 = np.array([[1,0],[0,1]])

train_data0 = np.random.multivariate_normal([0,0],cov0,N)
train_labels0 = np.zeros((N,1))
train_data1 = np.random.multivariate_normal([1,1],cov1,N)
train_labels1 = np.ones((N,1))
train_z = np.random.normal(0,1,N*2)
train_data1[:,1] = train_data1[:,1]+train_z[N:]
#train_z[:N] = 0
train_data = np.vstack([train_data0,train_data1])
train_labels = np.append(train_labels0,train_labels1)

# Plot data
plt.scatter(train_data[train_labels==0,0],train_data[train_labels==0,1],c="r",label='Y=0')
plt.scatter(train_data[train_labels==1,0],train_data[train_labels==1,1],c="b",label='Y=1')
plt.legend()
plt.show()

train_z = train_z.reshape(-1,1)

indices = np.random.permutation(len(train_data))
train_data = train_data[indices]
train_labels =train_labels[indices]
train_z = train_z[indices]

#train test split
train_data, valid_data, train_labels, valid_labels, train_z, valid_z = train_test_split(train_data,train_labels,train_z,test_size=50000)

#Models 
inputs = Input(shape=(train_data.shape[1],))

Dx = Dense(20, activation="tanh")(inputs)
Dx = Dense(20, activation="relu")(Dx)
Dx = Dense(1, activation="sigmoid")(Dx)
f = Model([inputs],[Dx])

from keras.layers.merge import concatenate, Add
out = f(inputs)
out = Dense(20,activation="relu")(out)
out = Dense(20,activation="relu")(out)
mu = Dense(5,activation="linear")(out)
sigma = Dense(5,activation="exponential")(out)
pi = Dense(5,activation="softmax")(out)
out = concatenate([mu,sigma,pi])
r = Model([inputs],[out]) 

# Loss for r
def rloss(lam):
  def loss_r(y_true,y_pred):
    y_true = y_true.ravel()
    mu = y_pred[:,:5]
    sig = y_pred[:,5:10]
    pi = y_pred[:,10:]
    pdf = 0
    for i in range(5):
      pdf += pi[:,i] *(1.0/(sig[:,i]*np.sqrt(2.0*np.pi)))*keras.backend.exp(-((y_true-mu[:,i])**2)/(2.0*(sig[:,i]**2)))
    return lam * keras.backend.mean(-keras.backend.log(pdf))
  return loss_r


f.compile(loss = keras.backend.binary_crossentropy, optimizer = keras.optimizers.SGD())

adv1 = Model([inputs], [f(inputs), r(inputs)])
adv1.compile(loss=[keras.backend.binary_crossentropy,rloss(-50)],optimizer=keras.optimizers.SGD(momentum=0.0))

adv2 = Model([inputs],[r(inputs)])
adv2.compile(loss=[rloss(1.0)],optimizer=keras.optimizers.SGD(momentum=0.0))

adv1.summary()
adv2.summary()

f.trainable = False
r.trainable = True
adv2.compile(loss=[rloss(1.0)],optimizer=keras.optimizers.SGD(momentum=0.0))
adv2.fit(train_data,train_z,epochs=20)
